tests/lib/rules/lines-between-class-methods.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * @fileoverview Tests for lines-between-class-methods rule.
+ * @author 薛定谔的猫<hh_2013@foxmail.com>
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+const rule = require("../../../lib/rules/lines-between-class-methods");
+const RuleTester = require("../../../lib/testers/rule-tester");
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+const ALWAYS_MESSAGE = "Methods must be padded by blank lines.";
+const NEVER_MESSAGE = "Methods must not be padded by blank lines.";
+
+//------------------------------------------------------------------------------
+// Tests
+//------------------------------------------------------------------------------
+
+const ruleTester = new RuleTester({ parserOptions: { ecmaVersion: 6 } });
+
+ruleTester.run("lines-between-class-methods", rule, {
+    valid: [
+        "class foo{}",
+        "class foo{\n\n}",
+        "class foo{constructor(){}\n}",
+        "class foo{\nconstructor(){}}",
+        "class foo{ bar(){}\n\nbaz(){}}",
+        { code: "class foo{ bar(){}\nbaz(){}}", options: ["never"] },
+        { code: "class foo{ bar(){}\n\nbaz(){}}", options: ["always"] },
+        { code: "class foo{ bar(){\n}\n\nbaz(){}}", options: [{ multiline: "always" }] },
+        { code: "class foo{ bar(){\n}\nbaz(){}}", options: [{ multiline: "never" }] },
+        { code: "class foo{ bar(){}\n\nbaz(){}}", options: [{ singleline: "always" }] },
+        { code: "class foo{ bar(){}\nbaz(){}}", options: [{ singleline: "never" }] }
```

Comment:
I'd like to see more tests, especially for semicolons and comments.

```js
class A {
	a() {}
    ;
	b() {}
}

class B {
	a() {}
    // comment
	b() {}
}

class C {
	a() {}
    /**
     * comment
     */
	b() {}
}
```

Reply:
> ```js
> class A {
>     a() {}
>     ;
>     b() {}
> }
> ```

should also add test for semi? it's so strange.

Reply:
> ```js
> class A {
>     a() {}
>     ;
>     b() {}
> }
> ```

should also add tests for semi? this is not a necessary and looks strange!

Reply:
Yeah, it's strange, but it's valid syntax. This rule should work well even if semicolons exist.

In semantics, [the semicolon is an empty member](https://www.ecma-international.org/ecma-262/8.0/#prod-ClassElement). So if this rule is checking blank lines between class members, this rule should check blank lines between a method and a semicolon, and between consecutive semicolons. But, this rule seems to check only between methods, so maybe it should ignore blank lines between a method and a semicolon, and between consecutive semicolons.

But I'm not sure if this rule should check only between methods. In this direction, we have to add rules as many as the combination of the number of class member types.

- lines-between-class-empty-members
- lines-between-class-methods
- lines-between-class-method-and-empty
- lines-between-class-fields (in future)
- lines-between-class-field-and-empty (in future)
- lines-between-class-field-and-method (in future)

I'm imaging a rule like [padding-line-between-statements](https://eslint.org/docs/rules/padding-line-between-statements) rule. Thought?

Reply:
👍 . a semicolon can be seen as method? if not, I think the rulename `lines-between-class-methods` is a little misleading.

Reply:
I think empty members are rare and not very useful, and it doesn't seem like people have a lot of different preferences for enforcing newlines between class members. So I'm wondering if it would be better to keep just `line-between-class-methods` for now, and evaluate proposals for other class members separately if people want more customizability.

Reply:
My opinion is not strongly. So I'm OK if the rule checks only methods and just ignores blank lines between a method and a semicolon, and between consecutive semicolons. But, I'm guessing `lines-between-class-methods` is going to be deprecated in future by adding fields.

Reply:
What if we call it `lines-between-class-members` instead? That way, we can add options to it for fields in the future.